expression/builtin_control.go

```diff
@@ -195,7 +194,7 @@ func (c *caseWhenFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 	}
 	fieldTp.Decimal, fieldTp.Flen = decimal, flen
 	if fieldTp.EvalType().IsStringKind() && !isBinaryStr {
-		fieldTp.Charset, fieldTp.Collate = charset.CharsetUTF8MB4, charset.CollationUTF8MB4
+		fieldTp.Charset, fieldTp.Collate = DeriveCollationFromExprs(ctx, args...)
```

Comment:
If it's binaryStr, the collation should be binary.

Reply:
Addressed. PTAL Again. @wjhuang2016